Reference/google/protobuf/compiler/plugin.pb.swift

```diff
@@ -79,7 +79,17 @@ public struct Google_Protobuf_Compiler_CodeGeneratorRequest: ProtobufGeneratedMe
   public var fileToGenerate: [String] = []
 
   ///   The generator parameter passed on the command-line.
-  public var parameter: String? = nil
+  private var _parameter: String? = nil
+  public var parameter: String {
+    get {return _parameter ?? ""}
+    set {_parameter = newValue}
+  }
+  public var hasParameter: Bool {
+    return _parameter != nil
+  }
+  public mutating func clearParameter() {
+    return _parameter = nil
+  }
```

Comment:
I'm still thinking about this.

## Arguments for this change

If I understand correctly, you have two key arguments for this change:

### 1. The spec comment about default values.

In particular, the proto2 spec says:
> If the default value is not specified for an optional element, a type-specific default value is used

I'm not especially persuaded by this:
 *  C++ and Java lack any general concept of an "unset property".  As a result, it's necessary for the spec to describe the value that should be seen if a property is not set (since properties in those languages can never be truly unset).
 * Python and Javascript do have a notion of an "unset property" but the syntax for testing them is awkward.
 * Swift `Optional` does provide a robust concept of an "unset variable" and provides trivial syntax (via '??' syntax) for working with possibly-unset values.

(Yes, I've weakened my own argument by attaching it to a `string` property here, since C++/Java reference types are nullable.)

### 2. Consistency with specified defaults

The concern here is whether these are truly the same:
```
    optional string parameter = 99;
```
and
```
    optional string parameter = 99 [default=""];
```
If you consider the first to just be shorthand for the second, then these are the same and we should generate the same results.  I suspect the most expert users of protobuf for C++ and Java do consider these to be the same and would be surprised if we generated different code for them.  Of course, such experts are unlikely to change one of these to the other.

## Arguments against this change

### 3. Generated code size

This change adds four methods (getter, setter, `has`, and `clear`) that are entirely unnecessary in the earlier form.

### 4. Explicit defaults are the less-common case

In helping protobuf users here, I've only seen a handful of uses of `default`.   Other examples: the proto2 language guide only shows two fields with explicit defaults (out of several dozen example fields),  descriptor.proto has 89 optional fields with only two defaults (both enum-valued).  The protobuf/src/google/protobuf has almost 1600 optional fields of which only 100 have explicit defaults.

### 5. Natural mapping of `optional` to `Optional`

Since `Optional` types are natural and well-supported in Swift, it seems odd to go to such lengths to avoid them.  The default values in the spec (zero, "", false, etc.) should not need to be explicitly provided by us.

So I fear that we're complicating the common case, generating code that will be confusing to Swift users, all in order to make an uncommon idiom a little easier to understand.